turbopack/crates/turbo-persistence/src/interning_serde/mod.rs

```diff
@@ -1,40 +1,89 @@
 //! Exposed for usage in `turbo-tasks-backend`
 
-use std::io::Write;
+use std::io::{Read, Write};
 
 use indexmap::IndexSet;
 use rustc_hash::FxBuildHasher;
-use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use serde::{de::DeserializeOwned, Serialize};
 use turbo_rcstr::RcStr;
 
-#[derive(Serialize, Deserialize)]
-struct Data(Vec<u8>, IndexSet<RcStr, FxBuildHasher>);
-
-pub fn to_vec<T>(config: &pot::Config, value: &T) -> pot::Result<Vec<u8>>
+pub fn to_vec<T>(config: &pot::Config, value: &T) -> anyhow::Result<(Vec<u8>, RcStrToLocalId)>
 where
     T: Serialize,
 {
     let mut vec = Vec::new();
-    to_writer(config, value, &mut vec)?;
-    Ok(vec)
+    let ser_map = to_writer(config, value, &mut vec)?;
+    Ok((vec, ser_map))
+}
+
+#[derive(Default)]
+pub struct RcStrToLocalId(pub IndexSet<RcStr, FxBuildHasher>);
+
+#[derive(Default)]
+pub struct LocalIdToGlobalId(Vec<u32>);
+
+impl From<Vec<u32>> for LocalIdToGlobalId {
+    fn from(value: Vec<u32>) -> Self {
+        Self(value)
+    }
+}
+
+impl LocalIdToGlobalId {
+    pub fn write_to(&self, writer: &mut impl Write) -> anyhow::Result<()> {
+        let len = self.0.len() as u32;
+        writer.write_all(&len.to_le_bytes())?;
+        for id in self.0.iter() {
+            writer.write_all(&id.to_le_bytes())?;
+        }
+        Ok(())
+    }
+
+    fn read_from(reader: &mut impl Read) -> anyhow::Result<Self> {
+        let mut global_ids = Vec::new();
+
+        let mut len = [0; 4];
+        reader.read_exact(&mut len)?;
+        let len = u32::from_le_bytes(len);
+        global_ids.reserve(len as usize);
+
+        for _ in 0..len {
+            let mut id = [0; 4];
+            reader.read_exact(&mut id)?;
+            global_ids.push(u32::from_le_bytes(id));
+        }
+
+        Ok(Self(global_ids))
+    }
 }
 
-pub fn to_writer<T, W>(config: &pot::Config, value: &T, writer: W) -> pot::Result<()>
+pub fn to_writer<T, W>(config: &pot::Config, value: &T, writer: W) -> anyhow::Result<RcStrToLocalId>
```

Comment:
I would expect `to_writer` and `from_slice` being the reversed version of each other.

Therefore, as `from_slice` completely handles `LocalIdToGlobalId`, I would expect `to_writer` to also take care of writing the `LocalIdToGlobalId` to the writer.

Technically this might be a bit challenging, but it will be easy only you convert the `RcStrToLocalId` in to a trailer instead of a header. So it stores:

```
[ serialized bytes ] [ global intern id ] [ global intern id ] ... N ... [ global intern id ] [ N ]
```

This way you can directly serialize to the writer and append the `RcStrToLocalId` struct.

in `from_slice` you need to grab the `N` value from the last 4 bytes and compute the start of the trailer.